File: kth-smallest-element-in-a-bst.js
// Time Complexity : O(H + k)
// Space Complexity : O(H) stack size
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */


// Recursive without array
let K,
    result;
var kthSmallest = function(root, k) {
    if (!root) return 0;
    
    K = k;
    
    inorder(root);
    return result;
};

var inorder = function(root) {
    if (root == null) return;
    
    inorder(root.left);
    K--;
    if (K == 0) {
        result = root.val;
        return;
    }
    inorder(root.right);
};

// Iterative without array
// var kthSmallest = function(root, k) {
//     if (!root) return 0;
    
//     const stack = [];
//     let cursor = root;
    
//     while (cursor != null) {
//         stack.push(cursor);
//         cursor = cursor.left;
//     }

//     while (stack.length > 0) {
//         cursor = stack.pop();
        
//         k--;
//         if (k == 0) return cursor.val;
        
//         if (cursor.right != null) {
//             cursor = cursor.right;
//             while (cursor != null) {
//                 stack.push(cursor) ;
//                 cursor = cursor.left;
//             }
//         }
//     }
    
//     return 0;
// };

// Recursive using an array
// var kthSmallest = function(root, k) {
//     if (!root) return 0;
//     let inorderTraversal = inorder(root, []);
//     return inorderTraversal[k - 1];
// };

// var inorder = function(root, arr) {
//     // Base Case
//     if (root == null) return arr;
//     // Logic
//     inorder(root.left, arr);
//     arr.push(root.val);
//     inorder(root.right, arr)
//     return arr
// }

File: src/KthSmallestElementinaBST.java
import java.util.Stack;

/*
 * The complexity of the algorithm is O(n) where n is no of elements.
 * Space complexity is O(k) where k is height of tree.
 */

public class KthSmallestElementinaBST {

	public class TreeNode {
		int val;
		TreeNode left;
		TreeNode right;

		TreeNode() {
		}

		TreeNode(int val) {
			this.val = val;
		}

		TreeNode(int val, TreeNode left, TreeNode right) {
			this.val = val;
			this.left = left;
			this.right = right;
		}
	}

	// Iterative approach

	public int kthSmallest(TreeNode root, int k) {

		int count = 0;
		int result = 0;
		Stack<TreeNode> st = new Stack<>();
		// st.push(root);

		while (!st.isEmpty() || root != null) {

			while (root != null) {
				st.push(root);
				root = root.left;

			}

			root = st.pop();

			count++;
			if (count == k) {
				result = root.val;
				break;
			}

			root = root.right;

		}

		return result;
	}

	// recursive

//  int result=0;
//  int count;
//  public int kthSmallest(TreeNode root, int k) {

//      inorder(root,k);
//      return result;
//  }

//  private void inorder(TreeNode root, int k){
//      // base
//      if(root==null)
//          return ;

//      //logic

//      inorder(root.left,k);
//      count++;
//      if(count == k){
//          result= root.val;
//          return;
//      }

//         if(result == 0)
//    			inorder(root.right,k);

//  }

	public static void main(String[] args) {
		// TODO Auto-generated method stub

	}

}
